hash_reader_test.go

```diff
@@ -0,0 +1,314 @@
+		t.Error("expected ionHashReader.Next() to return false")
+	} else {
+		err := ionHashReader.Err()
+		_, ok := err.(*UnknownSymbolError)
+		if !ok {
+			t.Error("expected ionHashReader.Next() to result in an UnknownSymbolError")
+		}
+	}
+}
+
+func TestIonReaderContract(t *testing.T) {
+	file, err := ioutil.ReadFile("ion_hash_tests.ion")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ionReader := ion.NewReaderBytes(file)
+
+	ionHashReader, err := NewHashReader(ionReader, NewCryptoHasherProvider(SHA256))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = Compare(ionReader, ionHashReader)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func ConsumeRemainderPartialConsume(ionHashReader HashReader) error {
+	ionHashReader.Next()
+	err := ionHashReader.StepIn()
+	if err != nil {
+		return err
+	}
+
+	ionHashReader.Next()
+	ionHashReader.Next()
+	ionHashReader.Next()
+	err = ionHashReader.StepIn()
+	if err != nil {
+		return err
+	}
+
+	ionHashReader.Next()
+	err = ionHashReader.StepOut() // we've only partially consumed the struct
+	if err != nil {
+		return err
+	}
+
+	err = ionHashReader.StepOut() // we've only partially consumed the list
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func ConsumeRemainderStepInStepOutNested(ionHashReader HashReader) error {
+	ionHashReader.Next()
+	err := ionHashReader.StepIn()
+	if err != nil {
+		return err
+	}
+
+	ionHashReader.Next()
+	ionHashReader.Next()
+	ionHashReader.Next()
+	err = ionHashReader.StepIn()
+	if err != nil {
+		return err
+	}
+
+	err = ionHashReader.StepIn()
+	if err != nil {
+		return err
+	}
+
+	err = ionHashReader.StepOut() // we haven't consumed ANY of the struct
+	if err != nil {
+		return err
+	}
+
+	err = ionHashReader.StepOut() // we've only partially consumed the list
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func ConsumeRemainderStepInNextStepOut(ionHashReader HashReader) error {
+	ionHashReader.Next()
+	err := ionHashReader.StepIn()
+	if err != nil {
+		return err
+	}
+
+	ionHashReader.Next()
+	err = ionHashReader.StepOut() // we've partially consumed the list
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func ConsumeRemainderStepInStepOutTopLevel(ionHashReader HashReader) error {
+	ionHashReader.Next()
+	sum, err := ionHashReader.Sum(nil)
+	if err != nil {
+		return err
+	}
+
+	if !reflect.DeepEqual(sum, []byte{}) {
+		return fmt.Errorf("sums don't match'")
+	}
+
+	err = ionHashReader.StepIn()
+	if err != nil {
+		return err
+	}
+
+	_, err = ionHashReader.Sum(nil)
+	if err != nil {
+		_, ok := err.(*InvalidOperationError)
+		if !ok {
+			return fmt.Errorf("expected ionHashReader.Sum(nil) to return an InvalidOperationError")
+		}
+	} else {
+		return fmt.Errorf("expected ionHashReader.Sum(nil) to return an error")
+	}
+
+	err = ionHashReader.StepOut() // we haven't consumed ANY of the list
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func ConsumeRemainderSingleNext(ionHashReader HashReader) error {
+	ionHashReader.Next()
+	ionHashReader.Next()
```

Comment:
function name implies single call to `Next()` implementation has two calls to `Next()`

Reply:
Done.